neurons/miner.py

```diff
@@ -364,20 +364,25 @@ async def run(self):
                 i = 0
 
                 # Use a fixed number of batches instead of cycling until window changes
-                # Calculate a reasonable default based on window size and batch processing time
                 window_seconds = self.hparams.blocks_per_window * 12  # Assuming 12s per block
-                estimated_batch_time = 1.0  # Estimated seconds per batch
-                max_batches = getattr(
-                    self.hparams, 
-                    'max_batches_per_window', 
-                    min(100, int(window_seconds * 0.7 / estimated_batch_time))  # Use 70% of window time
-                )
+                target_window_duration = 84  # Target 84s for window iteration - hardcoded since we can't modify hparams
+                # Reserve ~20s for non-training operations (loading, compression, gathering)
+                training_time_budget = 64  # seconds
+                estimated_batch_time = 0.5  # Start with 0.5s per batch estimate
+                max_batches = min(100, int(training_time_budget / estimated_batch_time))
+
+                # Track batch processing time to adjust dynamically
+                batch_start_time = time.time()
                 batch_count = 0
 
                 for batch in loader:  # No need for itertools.cycle
-                    if self.current_window != step_window or batch_count >= max_batches:
+                    # Check if we're approaching our time budget
+                    elapsed = time.time() - window_start
+                    if elapsed > (target_window_duration * 0.8) or batch_count >= max_batches or self.current_window != step_window:
+                        tplr.logger.info(f"Time budget approaching ({elapsed:.2f}s), stopping batch processing")
```

Comment:
I am not sure this works since you need to put at the beginning of the next window.